File: algorithms/c/recursive_example.py
def solve(m, Y, R, i):
    #print m, Y, R
    if m == 0:
        return True
    if i == len(Y):
        return False

    R[i] = not R[i]
    if solve(m-Y[i], Y, R, i+1):
        return True
    else:
        R[i] = not R[i]
        return solve(m, Y, R, i+1)


vectors = [[5, 7, 3, 1],
           [5, 7, 1, 3],
           [8, 2, 2, 1],
           [13, 1, 1, 1],
           [13, 0, 0, 0],
           [0, 10, 0, 3],
           [0, 10, 1, 3],

           [15, -5, 3, 4],
           [15, -5, 4, 3],
           ]

m = 13
for vec in vectors:
    mask = [0, ] * len(vec)
    if solve(m, vec, mask, 0):
        print vec, mask
    else:
        print 'fail', vec, mask

File: algorithms/c/recursive_example.pl
solve(0, [], []).
solve(0, [_V|Vs], [0|Ms]) :-
    solve(0, Vs, Ms).
solve(N, [], []) :-
    N \== 0,
    !, fail.
solve(Tot, [V|Vs], [M|Ms]) :-
    (M=0, solve(Tot, Vs, Ms))
    ;
    (M=1, Tot1 is Tot - V, solve(Tot1, Vs, Ms)).



tests([
       [5, 7, 3, 1],
       [5, 7, 3, 1],
       [5, 7, 1, 3],
       [8, 2, 2, 1],
       [13, 0, 0, 0],
       [0, 10, 0, 3],
       [15, -5, 3, 4],
       [15, -5, 4, 3],
       [1, 2, 3, 4]
      ]).

test(X) :-
    (solve(13, X, R) ->
        write(X), write(' '), write(R), write(' ok.'), nl
    ;
        write(X), write(' '), write(fail), nl
    ).

test_all([X|Xs]) :-
    test(X), test_all(Xs).
test_all([]).

test_all :-
    tests(Ts), test_all(Ts).
